In [ ]:
#@title
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
#@title
df = pd.read_csv('/content/drive/My Drive/Rhodes Python Projects/KFBS/KFBS FYXX Alumni Export Process.csv')

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning:

Columns (78,79) have mixed types.Specify dtype option on import or set low_memory=False.



In [ ]:
#@title
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33193 entries, 0 to 33192
Data columns (total 85 columns):
 #   Column                                                                              Non-Null Count  Dtype  
---  ------                                                                              --------------  -----  
 0   BUSINESSPROCESSOUTPUT_PKID                                                          33193 non-null  object 
 1   CONSTITUENTS_DECEASED                                                               33193 non-null  bool   
 2   CONSTITUENTS_ISINACTIVE                                                             33193 non-null  bool   
 3   CONSTITUENTS_LOOKUPID                                                               33193 non-null  int64  
 4   CONSTITUENTS_TITLECODEID_TRANSLATION                                                30821 non-null  object 
 5   CONSTITUENTS_Last Name                                                              33193 non-n

BUSINESSPROCESSOUTPUT_PKID  CONSTITUENTS_DECEASED  CONSTITUENTS_ISINACTIVE  \
0                     29,968                  False                    False   
1                      5,979                  False                    False   
2                     29,345                  False                    False   
3                     10,977                  False                    False   
4                     31,257                  False                    False   

   CONSTITUENTS_LOOKUPID CONSTITUENTS_TITLECODEID_TRANSLATION  \
0              812261759                                  NaN   
1              811677944                                  NaN   
2              399800029                                  NaN   
3              812265643                                  NaN   
4              699074813                                  NaN   

               CONSTITUENTS_Last Name CONSTITUENTS_FIRSTNAME  \
0  2015 Robert H. Slater Living Trust                    NaN   
1                   2GT Richmond Inc.                    NaN   
2                            2U, Inc.                    NaN   
3             36 Venture Capital, LLC                    NaN   
4               3M Company, Inc. - HQ                    NaN   

  CONSTITUENTS_MIDDLENAME          CONSTITUENTS_FORMATTEDNAME  \
0                     NaN  2015 Robert H. Slater Living Trust   
1                     NaN                   2GT Richmond Inc.   
2                     NaN                            2U, Inc.   
3                     NaN             36 Venture Capital, LLC   
4                     NaN               3M Company, Inc. - HQ   

  CONSTITUENTS_NICKNAME  ... CONSTITUENTS_EducationHistory1_3_DEGREE  \
0                   NaN  ...                                     NaN   
1                   NaN  ...                                     NaN   
2                   NaN  ...                                     NaN   
3                   NaN  ...                                     NaN   
4                   NaN  ...                                     NaN   

  CONSTITUENTS_EducationHistory1_3_ADDITIONALINFORMATION1_EDUCATIONALDEPARTMENTCODE  \
0                                                NaN                                  
1                                                NaN                                  
2                                                NaN                                  
3                                                NaN                                  
4                                                NaN                                  

  CONSTITUENTS_UNCBoardofVisitors_CONSTITUENT_NAME  \
0                                              NaN   
1                                              NaN   
2                                              NaN   
3                                              NaN   
4                                              NaN   

  CONSTITUENTS_UNCBoardofTrustees_CONSTITUENT1_NAME  \
0                                               NaN   
1                                               NaN   
2                                               NaN   
3                                               NaN   
4                                               NaN   

  CONSTITUENTS_UNCBoardofGovernors_CONSTITUENT2_NAME  \
0                                                NaN   
1                                                NaN   
2                                                NaN   
3                                                NaN   
4                                                NaN   

  CONSTITUENTS_Chancellor’sPhilanthropicCouncil_CONSTITUENT3_NAME  \
0                                                NaN                
1                                                NaN                
2                                                NaN                
3                                                NaN                
4                                                NaN        

The column names are hard to identify, and also mixed between upper- and lower-case letters. I want to simplify them below.

In [ ]:
#@title
df.columns = [x.lower() for x in df.columns]
df.columns = df.columns.str.replace(r'constituents_', '')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33193 entries, 0 to 33192
Data columns (total 85 columns):
 #   Column                                                                 Non-Null Count  Dtype  
---  ------                                                                 --------------  -----  
 0   businessprocessoutput_pkid                                             33193 non-null  object 
 1   deceased                                                               33193 non-null  bool   
 2   isinactive                                                             33193 non-null  bool   
 3   lookupid                                                               33193 non-null  int64  
 4   titlecodeid_translation                                                30821 non-null  object 
 5   last name                                                              33193 non-null  object 
 6   firstname                                                              31328 non-null 

I'm going to manually filter the columns I will not be using during this project.

In [ ]:
#@title
cols = [0, 1, 2, 3, 4, 5, 6, 7, 9, 10, 11, 13, 14, 15, 17, 18, 19, 20, 21, 22, 
        23, 24, 25, 26, 27, 28, 29, 30, 32, 33, 34, 36, 37, 41, 43, 44, 45, 48,  
        49, 50, 51, 52, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72,  
        73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84]
simple_df = df.drop(df.columns[cols], axis = 1)
simple_df.info()
simple_df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33193 entries, 0 to 33192
Data columns (total 17 columns):
 #   Column                                                      Non-Null Count  Dtype  
---  ------                                                      --------------  -----  
 0   formattedname                                               33193 non-null  object 
 1   uncdemographic_primaryconstituency                          33193 non-null  object 
 2   addressprimary_state                                        30044 non-null  object 
 3   educationhistory_degree                                     16933 non-null  object 
 4   educationhistory_2_degree                                   1067 non-null   object 
 5   prospect_unc ratings                                        18553 non-null  object 
 6   uncdemographic_unc_affinity_score                           26554 non-null  float64
 7   kfbssiteengagementscore_total_score                         30066 non-null  float64
 

formattedname uncdemographic_primaryconstituency  \
0  2015 Robert H. Slater Living Trust                              Trust   
1                   2GT Richmond Inc.                        Corporation   
2                            2U, Inc.                        Corporation   
3             36 Venture Capital, LLC                        Corporation   
4               3M Company, Inc. - HQ                        Corporation   

  addressprimary_state educationhistory_degree educationhistory_2_degree  \
0                   NC                     NaN                       NaN   
1                   VA                     NaN                       NaN   
2                   MD                     NaN                       NaN   
3                   TN                     NaN                       NaN   
4                   MN                     NaN                       NaN   

  prospect_unc ratings  uncdemographic_unc_affinity_score  \
0                  NaN                                NaN   
1                  NaN                                NaN   
2                  NaN                                NaN   
3                  NaN                                NaN   
4                  NaN                                NaN   

   kfbssiteengagementscore_total_score  uncdemographic1_household dedupe id  \
0                                  NaN                            812261759   
1                                  NaN                            811677944   
2                                  NaN                            399800029   
3                                  NaN                            812265643   
4                                  NaN                            699074813   

  uncaggregateacademicgivingalltime_unc all time recognition  \
0                                        $24,000.00            
1                                        $59,000.00            
2                                        $53,000.00            
3                                           $100.00            
4                                       $434,858.50            

  uncaggregateacademicgivingbysite_kfbs all time recognition kfbsfy22_amount  \
0                                        $24,000.00              $15,000.00    
1                                        $58,500.00              $50,000.00    
2                                        $53,000.00              $10,000.00    
3                                           $100.00                      NaN   
4                                         $5,000.00                      NaN   

  kfbsfy21_amount kfbsfy20_amount kfbsfy19_amount kfbsfy18_amount  \
0      $4,000.00       $5,000.00              NaN             NaN   
1             NaN             NaN      $8,000.00              NaN   
2             NaN             NaN     $10,000.00      $10,000.00    
3             NaN        $100.00              NaN             NaN   
4             NaN             NaN             NaN             NaN   

  kfbsfy17_amount  
0             NaN  
1             NaN  
2     $10,000.00   
3             NaN  
4             NaN

Further simplifying the column heads.

In [ ]:
#@title
simple_df.columns = simple_df.columns.str.replace(r'uncdemographic_', '')
simple_df.columns = simple_df.columns.str.replace(r'addressprimary_', '')
simple_df.columns = simple_df.columns.str.replace(r'educationhistory_', '')
simple_df.columns = simple_df.columns.str.replace(r'uncdemographic1_', '')
simple_df.columns = simple_df.columns.str.replace(r'uncaggregateacademicgivingalltime_', '')
simple_df.columns = simple_df.columns.str.replace(r'uncaggregateacademicgivingbysite_', '')

simple_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 33193 entries, 0 to 33192
Data columns (total 17 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   formattedname                        33193 non-null  object 
 1   primaryconstituency                  33193 non-null  object 
 2   state                                30044 non-null  object 
 3   degree                               16933 non-null  object 
 4   2_degree                             1067 non-null   object 
 5   prospect_unc ratings                 18553 non-null  object 
 6   unc_affinity_score                   26554 non-null  float64
 7   kfbssiteengagementscore_total_score  30066 non-null  float64
 8   household dedupe id                  33193 non-null  int64  
 9   unc all time recognition             33193 non-null  object 
 10  kfbs all time recognition            33193 non-null  object 
 11  kfbsfy22_amount             

In [ ]:
#@title
simple_df['degree'] = simple_df['degree'].fillna('Not Available')

The dollar amounts are in object data types. I want them in floats.

In [ ]:
#@title
simple_df['unc all time recognition'] = simple_df['unc all time recognition'].str.replace(',', '').str.replace('$', '').astype(float)
simple_df['kfbs all time recognition'] = simple_df['kfbs all time recognition'].str.replace(',', '').str.replace('$', '').astype(float)
simple_df['kfbsfy22_amount'] = simple_df['kfbsfy22_amount'].str.replace(',', '').str.replace('$', '').astype(float)
simple_df['kfbsfy21_amount'] = simple_df['kfbsfy21_amount'].str.replace(',', '').str.replace('$', '').astype(float)
simple_df['kfbsfy20_amount'] = simple_df['kfbsfy20_amount'].str.replace(',', '').str.replace('$', '').astype(float)
simple_df['kfbsfy19_amount'] = simple_df['kfbsfy19_amount'].str.replace(',', '').str.replace('$', '').astype(float)
simple_df['kfbsfy18_amount'] = simple_df['kfbsfy18_amount'].str.replace(',', '').str.replace('$', '').astype(float)
simple_df['kfbsfy17_amount'] = simple_df['kfbsfy17_amount'].str.replace(',', '').str.replace('$', '').astype(float)

simple_df.iloc[:, 9:17] = simple_df.iloc[:, 9:17].fillna(0)

simple_df.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: FutureWarning:

The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: FutureWarning:

The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: FutureWarning:

The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: FutureWarning:

The default value of regex will change from True to False in a future version. In addition, single character regular

formattedname primaryconstituency state  \
0  2015 Robert H. Slater Living Trust               Trust    NC   
1                   2GT Richmond Inc.         Corporation    VA   
2                            2U, Inc.         Corporation    MD   
3             36 Venture Capital, LLC         Corporation    TN   
4               3M Company, Inc. - HQ         Corporation    MN   

          degree 2_degree prospect_unc ratings  unc_affinity_score  \
0  Not Available      NaN                  NaN                 NaN   
1  Not Available      NaN                  NaN                 NaN   
2  Not Available      NaN                  NaN                 NaN   
3  Not Available      NaN                  NaN                 NaN   
4  Not Available      NaN                  NaN                 NaN   

   kfbssiteengagementscore_total_score  household dedupe id  \
0                                  NaN            812261759   
1                                  NaN            811677944   
2                                  NaN            399800029   
3                                  NaN            812265643   
4                                  NaN            699074813   

   unc all time recognition  kfbs all time recognition  kfbsfy22_amount  \
0                   24000.0                    24000.0          15000.0   
1                   59000.0                    58500.0          50000.0   
2                   53000.0                    53000.0          10000.0   
3                     100.0                      100.0              0.0   
4                  434858.5                     5000.0              0.0   

   kfbsfy21_amount  kfbsfy20_amount  kfbsfy19_amount  kfbsfy18_amount  \
0           4000.0           5000.0              0.0              0.0   
1              0.0              0.0           8000.0              0.0   
2              0.0              0.0          10000.0          10000.0   
3              0.0            100.0              0.0              0.0   
4              0.0              0.0              0.0              0.0   

   kfbsfy17_amount  
0              0.0  
1              0.0  
2          10000.0  
3              0.0  
4              0.0

To remove the household duplicates, I sorted the dataframe from `kfbs all time recognition`. 

In [ ]:
#@title
simple_df = simple_df.sort_values(by = ['kfbs all time recognition'], ascending = False)
simple_df.head()

formattedname        primaryconstituency state  \
31813              William R. Kenan Jr. Fund  Non-University Foundation    NC   
31812  William R. Kenan Jr. Charitable Trust                      Trust    NC   
2108                  Mr. Steven Durant Bell                    Alumnus    NC   
9804            Foundation for the Carolinas  Non-University Foundation    NC   
88             Ackerman, Charles (Estate of)             Alumnus Estate    GA   

              degree 2_degree                      prospect_unc ratings  \
31813  Not Available      NaN                                       NaN   
31812  Not Available      NaN                                       NaN   
2108   Not Available      NaN  AAA1 - $25,000,000-$49,999,999 Confident   
9804   Not Available      NaN                                       NaN   
88     Not Available      NaN                                       NaN   

       unc_affinity_score  kfbssiteengagementscore_total_score  \
31813                 NaN                                  NaN   
31812                 NaN                                  NaN   
2108                 70.0                                 84.0   
9804                  NaN                                  NaN   
88                    NaN                                  NaN   

       household dedupe id  unc all time recognition  \
31813            699007233              8.256471e+07   
31812            699005221              1.059756e+08   
2108             811206528              1.135435e+07   
9804             699002172              1.903032e+07   
88               812062105              8.456000e+06   

       kfbs all time recognition  kfbsfy22_amount  kfbsfy21_amount  \
31813                82124748.75       3120073.00       3867374.00   
31812                31019572.96        981966.00             0.00   
2108                 11228128.96       1245182.41       6192216.50   
9804                  8529610.33       2204533.44       1448522.65   
88                    8456000.00       3500000.00         56000.00   

       kfbsfy20_amount  kfbsfy19_amount  kfbsfy18_amount  kfbsfy17_amount  
31813       5366262.00       3034811.00        3050050.0        3009007.0  
31812        280000.00             0.00              0.0              0.0  
2108         588729.68         67000.00         200000.0              0.0  
9804        1295905.88       1325746.95        1119135.5         308850.0  
88          2800000.00       2100000.00              0.0              0.0

I also removed records with no donations in years 2017 through 2022, as well as constituents we do not want to analyze at this time.

In [ ]:
#@title
simple_df = simple_df.drop(simple_df[(simple_df.primaryconstituency =='Non-University Foundation') |
                                     (simple_df.primaryconstituency =='Family Foundation') |
                                     (simple_df.primaryconstituency =='Corporate Foundation') |
                                     (simple_df.primaryconstituency =='Corporation') |
                                     (simple_df.primaryconstituency =='Trust') |
                                     (simple_df.primaryconstituency =='Other Organization') |
                                     (simple_df.primaryconstituency =='Community Foundation') |
                                     (simple_df.primaryconstituency =='Former Faculty/Staff') |
                                     (simple_df.primaryconstituency =='Fundraising Consortium') |
                                     (simple_df.primaryconstituency =='Non-Alumnus Estate') |
                                     (simple_df.primaryconstituency =='Alumnus Estate')].index)


simple_df = simple_df.drop(simple_df[(simple_df.kfbsfy22_amount == 0) & 
                                     (simple_df.kfbsfy21_amount == 0) &
                                     (simple_df.kfbsfy20_amount == 0) &
                                     (simple_df.kfbsfy19_amount == 0) &
                                     (simple_df.kfbsfy18_amount == 0) &
                                     (simple_df.kfbsfy17_amount == 0)].index)

simple_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14322 entries, 2108 to 427
Data columns (total 17 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   formattedname                        14322 non-null  object 
 1   primaryconstituency                  14322 non-null  object 
 2   state                                14132 non-null  object 
 3   degree                               14322 non-null  object 
 4   2_degree                             571 non-null    object 
 5   prospect_unc ratings                 7962 non-null   object 
 6   unc_affinity_score                   14069 non-null  float64
 7   kfbssiteengagementscore_total_score  13970 non-null  float64
 8   household dedupe id                  14322 non-null  int64  
 9   unc all time recognition             14322 non-null  float64
 10  kfbs all time recognition            14322 non-null  float64
 11  kfbsfy22_amount            

Now, we can eliminate duplicates based on household number.

In [ ]:
#@title
simple_df = simple_df.drop_duplicates(subset = ['household dedupe id'])
simple_df.info()
simple_df.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10987 entries, 2108 to 427
Data columns (total 17 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   formattedname                        10987 non-null  object 
 1   primaryconstituency                  10987 non-null  object 
 2   state                                10820 non-null  object 
 3   degree                               10987 non-null  object 
 4   2_degree                             527 non-null    object 
 5   prospect_unc ratings                 5480 non-null   object 
 6   unc_affinity_score                   10764 non-null  float64
 7   kfbssiteengagementscore_total_score  10659 non-null  float64
 8   household dedupe id                  10987 non-null  int64  
 9   unc all time recognition             10987 non-null  float64
 10  kfbs all time recognition            10987 non-null  float64
 11  kfbsfy22_amount            

formattedname primaryconstituency state  \
2108        Mr. Steven Durant Bell             Alumnus    NC   
26439   Mr. William Gordon Seymour             Alumnus    NC   
30397         Mr. R. Steven Vetter             Alumnus    NC   
32397     Mr. Leonard William Wood             Alumnus    GA   
29646  Mr. John LeRoy Townsend III             Alumnus    CT   

                               degree 2_degree  \
2108                    Not Available      NaN   
26439      BS Business Administration      NaN   
30397      BS Business Administration      NaN   
32397  Master Business Administration      NaN   
29646  Master Business Administration      NaN   

                                    prospect_unc ratings  unc_affinity_score  \
2108            AAA1 - $25,000,000-$49,999,999 Confident                70.0   
26439               B1 - $1,000,000-$4,999,999 Confident                52.0   
30397           AAA1 - $25,000,000-$49,999,999 Confident                68.0   
32397               A1 - $5,000,000-$9,999,999 Confident                70.0   
29646  AAAA2 - $50,000,000-$99,999,999 Minimum Threshold                72.0   

       kfbssiteengagementscore_total_score  household dedupe id  \
2108                                  84.0            811206528   
26439                                 75.0            703842581   
30397                                 72.0            811238753   
32397                                 76.0            811303404   
29646                                 75.5            811230282   

       unc all time recognition  kfbs all time recognition  kfbsfy22_amount  \
2108                11354347.67                11228128.96       1245182.41   
26439                7602900.00                 7602625.00       2000000.00   
30397               14120052.03                 7060871.36       1000000.00   
32397               11449512.88                 5894612.88             0.00   
29646               11192632.29                 4964288.62             0.00   

       kfbsfy21_amount  kfbsfy20_amount  kfbsfy19_amount  kfbsfy18_amount  \
2108         6192216.5        588729.68          67000.0         200000.0   
26439        1250100.0       1250000.00        1250000.0        1000000.0   
30397        4455000.0         52446.36          10000.0         300000.0   
32397           5000.0          5000.00           5000.0         305000.0   
29646         540000.0        540589.18        1025000.0        1000000.0   

       kfbsfy17_amount  
2108               0.0  
26439         200000.0  
30397        1200000.0  
32397           5000.0  
29646          25000.0

Now we only have 10,987 records. We will add yearly change to the dataframe.

In [ ]:
#@title
fy17_18_df = simple_df[(simple_df['kfbsfy17_amount'] != 0) | (simple_df['kfbsfy18_amount'] != 0)]
fy17_18_df = fy17_18_df.drop(['kfbsfy19_amount', 'kfbsfy20_amount', 'kfbsfy21_amount', 'kfbsfy22_amount'], axis = 1)
fy17_18_df['changed_amount'] = fy17_18_df['kfbsfy18_amount'] - fy17_18_df['kfbsfy17_amount']
fy17_18_df = fy17_18_df.rename(columns = {'kfbsfy18_amount': 'curramt', 'kfbsfy17_amount': 'prevamt'})
fy17_18_df['fy'] = 'fy18'

fy18_19_df = simple_df[(simple_df['kfbsfy18_amount'] != 0) | (simple_df['kfbsfy19_amount'] != 0)]
fy18_19_df = fy18_19_df.drop(['kfbsfy17_amount', 'kfbsfy20_amount', 'kfbsfy21_amount', 'kfbsfy22_amount'], axis = 1)
fy18_19_df['changed_amount'] = fy18_19_df['kfbsfy19_amount'] - fy18_19_df['kfbsfy18_amount']
fy18_19_df = fy18_19_df.rename(columns = {'kfbsfy19_amount': 'curramt', 'kfbsfy18_amount': 'prevamt'})
fy18_19_df['fy'] = 'fy19'

fy19_20_df = simple_df[(simple_df['kfbsfy19_amount'] != 0) | (simple_df['kfbsfy20_amount'] != 0)]
fy19_20_df = fy19_20_df.drop(['kfbsfy17_amount', 'kfbsfy18_amount', 'kfbsfy21_amount', 'kfbsfy22_amount'], axis = 1)
fy19_20_df['changed_amount'] = fy19_20_df['kfbsfy20_amount'] - fy19_20_df['kfbsfy19_amount']
fy19_20_df = fy19_20_df.rename(columns = {'kfbsfy20_amount': 'curramt', 'kfbsfy19_amount': 'prevamt'})
fy19_20_df['fy'] = 'fy20'

fy20_21_df = simple_df[(simple_df['kfbsfy20_amount'] != 0) | (simple_df['kfbsfy21_amount'] != 0)]
fy20_21_df = fy20_21_df.drop(['kfbsfy17_amount', 'kfbsfy18_amount', 'kfbsfy19_amount', 'kfbsfy22_amount'], axis = 1)
fy20_21_df['changed_amount'] = fy20_21_df['kfbsfy21_amount'] - fy20_21_df['kfbsfy20_amount']
fy20_21_df = fy20_21_df.rename(columns = {'kfbsfy21_amount': 'curramt', 'kfbsfy20_amount': 'prevamt'})
fy20_21_df['fy'] = 'fy21'

fy21_22_df = simple_df[(simple_df['kfbsfy21_amount'] != 0) | (simple_df['kfbsfy22_amount'] != 0)]
fy21_22_df = fy21_22_df.drop(['kfbsfy17_amount', 'kfbsfy18_amount', 'kfbsfy19_amount', 'kfbsfy20_amount'], axis = 1)
fy21_22_df['changed_amount'] = fy21_22_df['kfbsfy22_amount'] - fy21_22_df['kfbsfy21_amount']
fy21_22_df = fy21_22_df.rename(columns = {'kfbsfy22_amount': 'curramt', 'kfbsfy21_amount': 'prevamt'})
fy21_22_df['fy'] = 'fy22'

In [ ]:
#@title
agg_df = pd.concat([fy17_18_df, fy18_19_df, fy19_20_df, fy20_21_df, fy21_22_df])

In [ ]:
#@title
conditions = [
              (agg_df['prevamt'] == 0) & (agg_df['curramt'] != 0),
              (agg_df['prevamt'] != 0) & (agg_df['curramt'] == 0),
              (agg_df['changed_amount'] < -10),
              (agg_df['changed_amount'] <= 10) & (agg_df['changed_amount'] >= -10),
              (agg_df['changed_amount'] > -10)]

values = ['new', 'lapsed', 'downgraded', 'same', 'upgraded']

agg_df['tier'] = np.select(conditions, values)
agg_df.info()
agg_df.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32326 entries, 2108 to 14021
Data columns (total 16 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   formattedname                        32326 non-null  object 
 1   primaryconstituency                  32326 non-null  object 
 2   state                                31972 non-null  object 
 3   degree                               32326 non-null  object 
 4   2_degree                             1687 non-null   object 
 5   prospect_unc ratings                 17935 non-null  object 
 6   unc_affinity_score                   31795 non-null  float64
 7   kfbssiteengagementscore_total_score  31692 non-null  float64
 8   household dedupe id                  32326 non-null  int64  
 9   unc all time recognition             32326 non-null  float64
 10  kfbs all time recognition            32326 non-null  float64
 11  curramt                  

formattedname primaryconstituency state  \
2108        Mr. Steven Durant Bell             Alumnus    NC   
26439   Mr. William Gordon Seymour             Alumnus    NC   
30397         Mr. R. Steven Vetter             Alumnus    NC   
32397     Mr. Leonard William Wood             Alumnus    GA   
29646  Mr. John LeRoy Townsend III             Alumnus    CT   

                               degree 2_degree  \
2108                    Not Available      NaN   
26439      BS Business Administration      NaN   
30397      BS Business Administration      NaN   
32397  Master Business Administration      NaN   
29646  Master Business Administration      NaN   

                                    prospect_unc ratings  unc_affinity_score  \
2108            AAA1 - $25,000,000-$49,999,999 Confident                70.0   
26439               B1 - $1,000,000-$4,999,999 Confident                52.0   
30397           AAA1 - $25,000,000-$49,999,999 Confident                68.0   
32397               A1 - $5,000,000-$9,999,999 Confident                70.0   
29646  AAAA2 - $50,000,000-$99,999,999 Minimum Threshold                72.0   

       kfbssiteengagementscore_total_score  household dedupe id  \
2108                                  84.0            811206528   
26439                                 75.0            703842581   
30397                                 72.0            811238753   
32397                                 76.0            811303404   
29646                                 75.5            811230282   

       unc all time recognition  kfbs all time recognition    curramt  \
2108                11354347.67                11228128.96   200000.0   
26439                7602900.00                 7602625.00  1000000.0   
30397               14120052.03                 7060871.36   300000.0   
32397               11449512.88                 5894612.88   305000.0   
29646               11192632.29                 4964288.62  1000000.0   

         prevamt  changed_amount    fy        tier  
2108         0.0        200000.0  fy18         new  
26439   200000.0        800000.0  fy18    upgraded  
30397  1200000.0       -900000.0  fy18  downgraded  
32397     5000.0        300000.0  fy18    upgraded  
29646    25000.0        975000.0  fy18    upgraded

# **All Donors**

## **Bar Charts**

In [ ]:
#@title
fig = px.histogram(agg_df, x = "tier",  barmode = 'group', text_auto = True, 
                   facet_col = 'fy', title = "All Donors by Tier by Fiscal Year", labels = {'fy': 'Fiscal Year'})
fig.update_xaxes(categoryorder = 'array',categoryarray = ['upgraded', 'downgraded', 'same', 'new', 'lapsed'])
fig.update_layout(yaxis_title = 'Number of Donors')
fig.show()

In [ ]:
#@title
fig = px.histogram(agg_df, x = "tier", color = 'primaryconstituency', barmode = 'stack', 
                   text_auto = True, facet_col = 'fy', title = "Constituency by Tier by Fiscal Year", 
                   labels = {'primaryconstituency': 'Primary Constituency',
                             'fy': 'Fiscal Year'})
fig.update_xaxes(categoryorder = 'array',categoryarray = ['upgraded', 'downgraded', 'same', 'new', 'lapsed'])
fig.update_layout(yaxis_title = 'Number of Donors')
fig.show()

In [ ]:
#@title
fig = px.histogram(agg_df, x = "tier", color = 'degree', barmode = 'stack', text_auto = True, 
                   facet_col = 'fy', title = "Degree by Tier by Fiscal Year",
                   labels = {'degree': 'Degree', 'fy': 'Fiscal Year'})
fig.update_xaxes(categoryorder = 'array',categoryarray = ['upgraded', 'downgraded', 'same', 'new', 'lapsed'])
fig.show()

In [ ]:
#@title
upgraded_df = agg_df[agg_df['tier'] == 'upgraded']
downgraded_df = agg_df[agg_df['tier'] == 'downgraded']
new_df = agg_df[agg_df['tier'] == 'new']
lapsed_df = agg_df[agg_df['tier'] == 'lapsed']
same_df = agg_df[agg_df['tier'] == 'same']

## **Line chart by constituency**

In [ ]:
#@title
upgraded_constituency_df = upgraded_df.groupby(['primaryconstituency', 'fy'])['formattedname'].count().reset_index()
upgraded_constituency_df = upgraded_constituency_df.rename(columns = {'formattedname': 'constituency count'})

new_constituency_df = new_df.groupby(['primaryconstituency', 'fy'])['formattedname'].count().reset_index()
new_constituency_df = new_constituency_df.rename(columns = {'formattedname': 'constituency count'})

downgraded_constituency_df = downgraded_df.groupby(['primaryconstituency', 'fy'])['formattedname'].count().reset_index()
downgraded_constituency_df = downgraded_constituency_df.rename(columns = {'formattedname': 'constituency count'})

lapsed_constituency_df = lapsed_df.groupby(['primaryconstituency', 'fy'])['formattedname'].count().reset_index()
lapsed_constituency_df = lapsed_constituency_df.rename(columns = {'formattedname': 'constituency count'})

same_constituency_df = same_df.groupby(['primaryconstituency', 'fy'])['formattedname'].count().reset_index()
same_constituency_df = same_constituency_df.rename(columns = {'formattedname': 'constituency count'})

In [ ]:
#@title
fig = px.line(upgraded_constituency_df, x = 'fy', y = 'constituency count', color = 'primaryconstituency', 
              text = 'constituency count', title = 'Donors who Upgraded by Constituency', 
              labels = {'primaryconstituency': 'Primary Constituency', 'fy': 'Fiscal Year', 'constituency count': 'Constituency Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

In [ ]:
#@title
fig = px.line(downgraded_constituency_df, x = 'fy', y = 'constituency count', color = 'primaryconstituency', 
              text = 'constituency count', title = 'Donors who Downgraded by Constituency', 
              labels = {'primaryconstituency': 'Primary Constituency', 'fy': 'Fiscal Year', 'constituency count': 'Constituency Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

In [ ]:
#@title
fig = px.line(same_constituency_df, x = 'fy', y = 'constituency count', color = 'primaryconstituency', 
              text = 'constituency count', title = 'Donors who Stayed the Same by Constituency', 
              labels = {'primaryconstituency': 'Primary Constituency', 'fy': 'Fiscal Year', 'constituency count': 'Constituency Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

In [ ]:
#@title
fig = px.line(new_constituency_df, x = 'fy', y = 'constituency count', color = 'primaryconstituency', 
              text = 'constituency count', title = 'New Donors by Constituency',
              labels = {'primaryconstituency': 'Primary Constituency', 'fy': 'Fiscal Year', 'constituency count': 'Constituency Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

In [ ]:
#@title
fig = px.line(lapsed_constituency_df, x = 'fy', y = 'constituency count', color = 'primaryconstituency', 
              text = 'constituency count', title = 'Lapsed Donors by Constituency',
              labels = {'primaryconstituency': 'Primary Constituency', 'fy': 'Fiscal Year', 'constituency count': 'Constituency Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

## **Line chart by degree**

In [ ]:
#@title
upgraded_degree_df = upgraded_df.groupby(['degree', 'fy'])['formattedname'].count().reset_index()
upgraded_degree_df = upgraded_degree_df.rename(columns = {'formattedname': 'degree count'})

new_degree_df = new_df.groupby(['degree', 'fy'])['formattedname'].count().reset_index()
new_degree_df = new_degree_df.rename(columns = {'formattedname': 'degree count'})

downgraded_degree_df = downgraded_df.groupby(['degree', 'fy'])['formattedname'].count().reset_index()
downgraded_degree_df = downgraded_degree_df.rename(columns = {'formattedname': 'degree count'})

lapsed_degree_df = lapsed_df.groupby(['degree', 'fy'])['formattedname'].count().reset_index()
lapsed_degree_df = lapsed_degree_df.rename(columns = {'formattedname': 'degree count'})

same_degree_df = same_df.groupby(['degree', 'fy'])['formattedname'].count().reset_index()
same_degree_df = same_degree_df.rename(columns = {'formattedname': 'degree count'})

In [ ]:
#@title
fig = px.line(upgraded_degree_df, x = 'fy', y = 'degree count', color = 'degree', 
              text = 'degree count', title = 'Donors who Upgraded by Degree',
              labels = {'degree': 'Degree', 'fy': 'Fiscal Year', 'degree count': 'Degree Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

In [ ]:
#@title
fig = px.line(downgraded_degree_df, x = 'fy', y = 'degree count', color = 'degree', 
              text = 'degree count', title = 'Donors who Downgraded by Degree',
              labels = {'degree': 'Degree', 'fy': 'Fiscal Year', 'degree count': 'Degree Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

In [ ]:
#@title
fig = px.line(same_degree_df, x = 'fy', y = 'degree count', color = 'degree', 
              text = 'degree count', title = 'Donors who Stayed the Same by Degree',
              labels = {'degree': 'Degree', 'fy': 'Fiscal Year', 'degree count': 'Degree Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

In [ ]:
#@title
fig = px.line(new_degree_df, x = 'fy', y = 'degree count', color = 'degree', 
              text = 'degree count', title = 'New Donors by Degree',
              labels = {'degree': 'Degree', 'fy': 'Fiscal Year', 'degree count': 'Degree Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

In [ ]:
#@title
fig = px.line(lapsed_degree_df, x = 'fy', y = 'degree count', color = 'degree', 
              text = 'degree count', title = 'Lapsed Donors by Degree',
              labels = {'degree': 'Degree', 'fy': 'Fiscal Year', 'degree count': 'Degree Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

# **Leadership Giving** 

In [ ]:
#@title
leadership_df = agg_df[(agg_df['curramt'] >= 2000) | (agg_df['prevamt'] >= 2000)]
leadership_df.info()
leadership_df.tail()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4535 entries, 2108 to 7874
Data columns (total 16 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   formattedname                        4535 non-null   object 
 1   primaryconstituency                  4535 non-null   object 
 2   state                                4516 non-null   object 
 3   degree                               4535 non-null   object 
 4   2_degree                             339 non-null    object 
 5   prospect_unc ratings                 3945 non-null   object 
 6   unc_affinity_score                   4440 non-null   float64
 7   kfbssiteengagementscore_total_score  4438 non-null   float64
 8   household dedupe id                  4535 non-null   int64  
 9   unc all time recognition             4535 non-null   float64
 10  kfbs all time recognition            4535 non-null   float64
 11  curramt                    

formattedname     primaryconstituency state  \
9311       Ms. Rose Shen Feng                 Alumnus    CA   
23952      Ms. Clare Quadland  Parent-Current Student    NC   
14345         Ms. Mary Ingram  Parent-Current Student    NC   
15233      Ms. Heather Joyner  Parent-Current Student    NC   
7874   Daniel David Domeracki                  Friend    NC   

                           degree 2_degree  \
9311   BS Business Administration      NaN   
23952               Not Available      NaN   
14345               Not Available      NaN   
15233               Not Available      NaN   
7874                Not Available      NaN   

                             prospect_unc ratings  unc_affinity_score  \
9311     C2 - $500,000-$999,999 Minimum Threshold                16.0   
23952  E3 - $25,000-$99,999 Unconfirmed/Screening                24.0   
14345  E3 - $25,000-$99,999 Unconfirmed/Screening                26.0   
15233  E3 - $25,000-$99,999 Unconfirmed/Screening                22.0   
7874                                          NaN                10.0   

       kfbssiteengagementscore_total_score  household dedupe id  \
9311                                  36.0            812261814   
23952                                 56.8            812213932   
14345                                 48.8            812215128   
15233                                 43.0            812213648   
7874                                  30.0            812393794   

       unc all time recognition  kfbs all time recognition  curramt  prevamt  \
9311                     2000.0                     2000.0   2000.0      0.0   
23952                    2000.0                     2000.0   2000.0      0.0   
14345                    2880.0                     2000.0   2000.0      0.0   
15233                    4500.0                     2000.0   2000.0      0.0   
7874                     2000.0                     2000.0      0.0   2000.0   

       changed_amount    fy    tier  
9311           2000.0  fy22     new  
23952          2000.0  fy22     new  
14345          2000.0  fy22     new  
15233          2000.0  fy22     new  
7874          -2000.0  fy22  lapsed

## **Bar Charts**

In [ ]:
#@title
fig = px.histogram(leadership_df, x = "tier",  barmode = 'group', text_auto = True, 
                   facet_col = 'fy', title = "Leadership Giving Donors by Tier by Fiscal Year", labels = {'fy': 'Fiscal Year'})
fig.update_xaxes(categoryorder = 'array',categoryarray = ['upgraded', 'downgraded', 'same', 'new', 'lapsed'])
fig.update_layout(yaxis_title = 'Number of Donors')
fig.show()

In [ ]:
#@title
fig = px.histogram(leadership_df, x = "tier", color = 'primaryconstituency', barmode = 'stack', 
                   text_auto = True, facet_col = 'fy', title = "Constituency by Tier by Fiscal Year", 
                   labels = {'primaryconstituency': 'Primary Constituency',
                             'fy': 'Fiscal Year'})
fig.update_xaxes(categoryorder = 'array',categoryarray = ['upgraded', 'downgraded', 'same', 'new', 'lapsed'])
fig.update_layout(yaxis_title = 'Number of Donors')
fig.show()

In [ ]:
#@title
fig = px.histogram(leadership_df, x = "tier", color = 'degree', barmode = 'stack', text_auto = True, 
                   facet_col = 'fy', title = "Degree by Tier by Fiscal Year",
                   labels = {'degree': 'Degree', 'fy': 'Fiscal Year'})
fig.update_xaxes(categoryorder = 'array',categoryarray = ['upgraded', 'downgraded', 'same', 'new', 'lapsed'])
fig.show()

In [ ]:
#@title
upgraded_df = leadership_df[leadership_df['tier'] == 'upgraded']
downgraded_df = leadership_df[leadership_df['tier'] == 'downgraded']
new_df = leadership_df[leadership_df['tier'] == 'new']
lapsed_df = leadership_df[leadership_df['tier'] == 'lapsed']
same_df = leadership_df[leadership_df['tier'] == 'same']

## **Line chart by constituency**

In [ ]:
#@title
upgraded_constituency_df = upgraded_df.groupby(['primaryconstituency', 'fy'])['formattedname'].count().reset_index()
upgraded_constituency_df = upgraded_constituency_df.rename(columns = {'formattedname': 'constituency count'})

new_constituency_df = new_df.groupby(['primaryconstituency', 'fy'])['formattedname'].count().reset_index()
new_constituency_df = new_constituency_df.rename(columns = {'formattedname': 'constituency count'})

downgraded_constituency_df = downgraded_df.groupby(['primaryconstituency', 'fy'])['formattedname'].count().reset_index()
downgraded_constituency_df = downgraded_constituency_df.rename(columns = {'formattedname': 'constituency count'})

lapsed_constituency_df = lapsed_df.groupby(['primaryconstituency', 'fy'])['formattedname'].count().reset_index()
lapsed_constituency_df = lapsed_constituency_df.rename(columns = {'formattedname': 'constituency count'})

same_constituency_df = same_df.groupby(['primaryconstituency', 'fy'])['formattedname'].count().reset_index()
same_constituency_df = same_constituency_df.rename(columns = {'formattedname': 'constituency count'})

In [ ]:
#@title
fig = px.line(upgraded_constituency_df, x = 'fy', y = 'constituency count', color = 'primaryconstituency', 
              text = 'constituency count', title = 'Donors who Upgraded by Constituency', 
              labels = {'primaryconstituency': 'Primary Constituency', 'fy': 'Fiscal Year', 'constituency count': 'Constituency Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

In [ ]:
#@title
fig = px.line(downgraded_constituency_df, x = 'fy', y = 'constituency count', color = 'primaryconstituency', 
              text = 'constituency count', title = 'Donors who Downgraded by Constituency', 
              labels = {'primaryconstituency': 'Primary Constituency', 'fy': 'Fiscal Year', 'constituency count': 'Constituency Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

In [ ]:
#@title
fig = px.line(same_constituency_df, x = 'fy', y = 'constituency count', color = 'primaryconstituency', 
              text = 'constituency count', title = 'Donors who Stayed the Same by Constituency', 
              labels = {'primaryconstituency': 'Primary Constituency', 'fy': 'Fiscal Year', 'constituency count': 'Constituency Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

In [ ]:
#@title
fig = px.line(new_constituency_df, x = 'fy', y = 'constituency count', color = 'primaryconstituency', 
              text = 'constituency count', title = 'New Donors by Constituency',
              labels = {'primaryconstituency': 'Primary Constituency', 'fy': 'Fiscal Year', 'constituency count': 'Constituency Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

In [ ]:
#@title
fig = px.line(lapsed_constituency_df, x = 'fy', y = 'constituency count', color = 'primaryconstituency', 
              text = 'constituency count', title = 'Lapsed Donors by Constituency',
              labels = {'primaryconstituency': 'Primary Constituency', 'fy': 'Fiscal Year', 'constituency count': 'Constituency Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

## **Line chart by degree**

In [ ]:
#@title
upgraded_degree_df = upgraded_df.groupby(['degree', 'fy'])['formattedname'].count().reset_index()
upgraded_degree_df = upgraded_degree_df.rename(columns = {'formattedname': 'degree count'})

new_degree_df = new_df.groupby(['degree', 'fy'])['formattedname'].count().reset_index()
new_degree_df = new_degree_df.rename(columns = {'formattedname': 'degree count'})

downgraded_degree_df = downgraded_df.groupby(['degree', 'fy'])['formattedname'].count().reset_index()
downgraded_degree_df = downgraded_degree_df.rename(columns = {'formattedname': 'degree count'})

lapsed_degree_df = lapsed_df.groupby(['degree', 'fy'])['formattedname'].count().reset_index()
lapsed_degree_df = lapsed_degree_df.rename(columns = {'formattedname': 'degree count'})

same_degree_df = same_df.groupby(['degree', 'fy'])['formattedname'].count().reset_index()
same_degree_df = same_degree_df.rename(columns = {'formattedname': 'degree count'})

In [ ]:
#@title
fig = px.line(upgraded_degree_df, x = 'fy', y = 'degree count', color = 'degree', 
              text = 'degree count', title = 'Donors who Upgraded by Degree',
              labels = {'degree': 'Degree', 'fy': 'Fiscal Year', 'degree count': 'Degree Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

In [ ]:
#@title
fig = px.line(downgraded_degree_df, x = 'fy', y = 'degree count', color = 'degree', 
              text = 'degree count', title = 'Donors who Downgraded by Degree',
              labels = {'degree': 'Degree', 'fy': 'Fiscal Year', 'degree count': 'Degree Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

In [ ]:
#@title
fig = px.line(same_degree_df, x = 'fy', y = 'degree count', color = 'degree', 
              text = 'degree count', title = 'Donors who Stayed the Same by Degree',
              labels = {'degree': 'Degree', 'fy': 'Fiscal Year', 'degree count': 'Degree Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

In [ ]:
#@title
fig = px.line(new_degree_df, x = 'fy', y = 'degree count', color = 'degree', 
              text = 'degree count', title = 'New Donors by Degree',
              labels = {'degree': 'Degree', 'fy': 'Fiscal Year', 'degree count': 'Degree Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()

In [ ]:
#@title
fig = px.line(lapsed_degree_df, x = 'fy', y = 'degree count', color = 'degree', 
              text = 'degree count', title = 'Lapsed Donors by Degree',
              labels = {'degree': 'Degree', 'fy': 'Fiscal Year', 'degree count': 'Degree Count'})
fig.update_traces(textposition = 'bottom center')
fig.update_xaxes(dtick=1)
fig.show()